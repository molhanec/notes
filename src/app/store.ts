import { Action, configureStore, getDefaultMiddleware, ThunkAction } from '@reduxjs/toolkit'
import { connectRouter, routerMiddleware } from 'connected-react-router'
import { createBrowserHistory } from 'history'

import counterReducer from '../features/counter/counterSlice'
import notesReducer from '../features/notes/notesSlice'

export const history = createBrowserHistory()

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    router: connectRouter(history) as any,
    notes: notesReducer,
  },
  middleware: [...getDefaultMiddleware(), routerMiddleware(history)],
})

export type RootState = ReturnType<typeof store.getState>
export type AppThunk = ThunkAction<void, RootState, unknown, Action<string>>
