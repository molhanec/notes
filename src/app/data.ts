import { AppThunk } from "./store"
import { loadNotesAsync } from "../features/noteList/notesSlice"
import { setProgress, setSuccess, setError } from "../features/info/infoSlice"

const rootDataUrl =
  "https://private-anon-3ab0da907c-note10.apiary-mock.com/notes"

const checkResponseStatus = (response: Response) => {
  if (response.ok) {
    return response
  } else {
    return Promise.reject(
      `Error fetching data: ${response.status} ${response.statusText}`
    )
  }
}

export const fetchJson = <T>(path: string) =>
  fetch(rootDataUrl + path)
    .then(checkResponseStatus)
    .then(response => response!.json() as Promise<T>)

export const addNoteAsync = (title: string): AppThunk => async dispatch => {
  dispatch(setProgress())
  try {
    fetch(rootDataUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        title,
      }),
    }).then(checkResponseStatus)

    dispatch(setSuccess("New note added"))
    dispatch(loadNotesAsync())
  } catch {
    dispatch(setError("Cannot add new note"))
  }
}
