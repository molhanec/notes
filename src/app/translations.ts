import { useSelector } from "react-redux"
import { RootState } from "./store"

export type LanguageCodes = "cz" | "en"
export const defaultLanguage = "cz"

interface Phrases {
  appTitle: string
  noteLoadFailure: string
  notesLoadFailure: string

  noteCreateSuccess: string
  noteCreateFailure: string
  noteUpdateSuccess: string
  noteUpdateFailure: string
  noteDeleteSuccess: string
  noteDeleteFailure: string

  formTitle: string
  formSave: string
  formDelete: string

  editFormHeader: string
  newFormHeader: string
}

export type PhraseKeys = keyof Phrases

const enPhrases: Phrases = {
  appTitle: "Notes",
  noteLoadFailure: "Cannot load note",
  notesLoadFailure: "Cannot load notes",

  noteCreateSuccess: "New note added",
  noteCreateFailure: "Cannot add new note",
  noteUpdateSuccess: "Note updated",
  noteUpdateFailure: "Cannot update note",
  noteDeleteSuccess: "Note deleted",
  noteDeleteFailure: "Cannot delete note",

  formTitle: "Title",
  formSave: "Save",
  formDelete: "Delete",

  editFormHeader: "Note details",
  newFormHeader: "Create new note",
}

const czPhrases: Phrases = {
  appTitle: "Poznámky",
  noteLoadFailure: "Nelze načíst poznámku",
  notesLoadFailure: "Nelze načíst poznámky",

  noteCreateSuccess: "Přidána poznámka",
  noteCreateFailure: "Nelze přidat poznámku",
  noteUpdateSuccess: "Poznámka upravena",
  noteUpdateFailure: "Nelze upravit poznámku",
  noteDeleteSuccess: "Poznámka smazána",
  noteDeleteFailure: "Nelze smazat poznámku",

  formTitle: "Nadpis",
  formSave: "Uložit",
  formDelete: "Smazat",

  editFormHeader: "Detaily poznámky",
  newFormHeader: "Vytvořit novou poznámku",
}

export const getPhrase = (language: "cz" | "en", key: PhraseKeys): string => {
  switch (language) {
    case "cz":
      return czPhrases[key]
    case "en":
      return enPhrases[key]
  }
}

export const useTranslate = () => {
  const language = useSelector((state: RootState) => state.language.code)
  return (key: PhraseKeys) => getPhrase(language, key)
}
