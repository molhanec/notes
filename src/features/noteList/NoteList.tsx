import React, { useEffect } from "react"
import { connect, ResolveThunks } from "react-redux"

import { RootState } from "../../app/store"
import { loadNotesAsync } from "./notesSlice"
import { setNoteDetails } from "../noteDetails/noteDetailsSlice"
import { Card } from "reactstrap"
import styled from "styled-components"
import { Link } from "react-router-dom"

const StyledLink = styled(Link)`
  color: white;
  &:hover {
    color: white;
    text-decoration: none;
  }
`

const StyledCard = styled(Card)`
  &:hover {
    box-shadow: 0px 0px 10px 2px rgba(0, 0, 0, 0.75);
  }
`

type Props = ReturnType<typeof mapStateToProps> &
  ResolveThunks<typeof mapDispatchToProps>

const NoteList: React.FC<Props> = ({
  notes,
  loadNotesAsync,
  setNoteDetails,
}: Props) => {
  useEffect(() => {
    setNoteDetails(undefined)
    loadNotesAsync()
  }, [loadNotesAsync, setNoteDetails])

  return (
    <>
      {notes.map(note => (
        <StyledCard key={note.id} body color="info" className="mb-3">
          <StyledLink to={`/note/${note.id}`} className="stretched-link">
            {note.title}
          </StyledLink>
        </StyledCard>
      ))}
    </>
  )
}

const mapStateToProps = (state: RootState) => ({ notes: state.notes.notes })

const mapDispatchToProps = {
  loadNotesAsync,
  setNoteDetails,
}

const ConnectedNoteList = connect(mapStateToProps, mapDispatchToProps)(NoteList)
export default ConnectedNoteList
