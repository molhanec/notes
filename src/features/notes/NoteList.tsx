import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'

import { RootState } from '../../app/store'
import { fetchJson } from '../../utils/data'
import { Notes, setNotes } from './notesSlice'

interface Props {
  notes: Notes
  setNotes: (notes: Notes) => void
}

const NoteList = ({ notes, setNotes }: Props) => {
  const [error, setError] = useState<boolean>(false)

  useEffect(() => {
    fetchJson<Notes>(
      "https://private-anon-3ab0da907c-note10.apiary-mock.com/notes"
    )
      .then(data => setNotes(data))
      .catch(error => setError(true))
  }, [setNotes])

  return (
    <>
      {error && <div>Cannot get the data!</div>}
      {notes.map(note => (
        <div key={note.id}>{note.title}</div>
      ))}
    </>
  )
}

const mapStateToProps = (state: RootState) => ({ notes: state.notes.notes })

const mapDispatchToProps = {
  setNotes,
}

const ConnectedNoteList = connect(mapStateToProps, mapDispatchToProps)(NoteList)
export default ConnectedNoteList
