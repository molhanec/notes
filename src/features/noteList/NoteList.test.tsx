import React from "react"
import { render } from "@testing-library/react"
import { Provider } from "react-redux"
import { store } from "../../app/store"
import { NoteList } from "./NoteList"
import { Router } from "react-router-dom"
import { createMemoryHistory } from "history"

test("<NoteList> matches snapshot", () => {
  const loadNotesAsync = jest.fn()
  const setNoteDetails = jest.fn()
  const { container } = render(
    <Provider store={store}>
      <Router history={createMemoryHistory()}>
        <NoteList
          notes={[
            {
              id: 1,
              title: "title 1",
            },
            {
              id: 2,
              title: "title 2",
            },
          ]}
          loadNotesAsync={loadNotesAsync}
          setNoteDetails={setNoteDetails as any}
        />
      </Router>
    </Provider>
  )

  expect(container).toMatchSnapshot()
  expect(loadNotesAsync).toBeCalled()
  expect(setNoteDetails).toBeCalledWith(undefined)
})
