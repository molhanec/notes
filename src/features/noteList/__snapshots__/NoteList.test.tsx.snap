// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteList> matches snapshot 1`] = `
<div>
  <div
    class="sc-AxirZ diIALt mb-3 card card-body bg-info"
  >
    <a
      class="sc-AxjAm cgUheY stretched-link"
      href="/note/1"
    >
      title 1
    </a>
  </div>
  <div
    class="sc-AxirZ diIALt mb-3 card card-body bg-info"
  >
    <a
      class="sc-AxjAm cgUheY stretched-link"
      href="/note/2"
    >
      title 2
    </a>
  </div>
</div>
`;
